storage/innobase/log/log0log.cc

```diff
@@ -87,6 +87,98 @@ void log_t::set_capacity()
 }
 
 #ifdef HAVE_PMEM
+# ifdef __x86_64__
+void (*log_t::pmem_persister)(const void *, size_t);
+
+static void log_clflush(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__("clflush %0"::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+}
+
+static void log_clflushopt(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__(".byte 0x66; clflush %0" /* clflushopt */::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+  __asm__ __volatile__("sfence":::"memory");
+}
+
+static void log_clwb(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__(".byte 0x66; xsaveopt %0" /* clwb */::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+  __asm__ __volatile__("sfence":::"memory");
+}
+
+#  include <cpuid.h>
+inline void log_t::pmem_persister_init()
+{
+  uint32_t eax= 0, ebx= 0, ecx= 0, edx= 0;
+  __cpuid(7, eax, ebx, ecx, edx);
+  if (ebx & 1U<<24 /* CLWB */)
+    pmem_persister= log_clwb;
+  else if (ebx & 1U<<23 /* CLFLUSHOPT */)
+    pmem_persister= log_clflushopt;
+  else
+    pmem_persister= log_clflush;
+}
+# elif defined __aarch64__
+void (*log_t::pmem_persister)(const void *, size_t);
+
+static void log_cvac(const void* buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__("dc cvac, %0" :: "r"(u) : "memory");
+  __asm__ __volatile__("dmb ishst" ::: "memory");
+}
+
+static void log_cvap(const void* buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__(".arch armv8.2-a\n dc cvap, %0"::"r"(u):"memory");
+  __asm__ __volatile__("dmb ishst" ::: "memory");
+}
+
+#  include <sys/auxv.h>
+#  include <asm/hwcap.h>
+#  ifndef HWCAP_DCPOP
+#   define HWCAP_DCPOP (1 << 16)
+#  endif
+
+inline void log_t::pmem_persister_init()
+{
+  pmem_persister= (getauxval(AT_HWCAP) & HWCAP_DCPOP) ? log_cvap : log_cvac;
+}
+# else
+void log_t::pmem_persist(const void *buf, size_t size)
+{
+#  ifdef __ppc64__
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__("dcbf 0,%0,6"::r(u):"memory");
```

Comment:
Based on tests on https://godbolt.org and its "compile to binary code" option and some reading of the LLVM revision history I came up with the following:
```c++
#   if __GNUC__ >= 11 || (defined __clang_major__ && __clang_major__ >= 12)
    __asm__ __volatile__("dcbstps 0,%0"::r(u):"memory");
#   else
    __asm__ __volatile__(".long (0x7cc000AC | %0 << 11)" :: "r"(u):"memory");
#   endif
```
I hope that the GCC version check is an adequate proxy for having a recent enough assembler. In clang, the assembler is built-in. The GCC source code itself does not contain either mnemonic `dcbstps` or `phwsync`.